### PR TITLE
hackage2nix: Split configuration, auto disable hydra builds

### PR DESCRIPTION
**Pull request by maralorn**
###### Motivation for this change

This is the change mentioned in https://github.com/NixOS/cabal2nix/pull/494.

It’s still a draft and needs to be adapted after #494 is merged. (This concerns especially the regenerate-hackage-packages script, which is patched to do local things right now.)

Since we have multiple scripts editing hackage2nix config and users also doing that, we decided it would be prudent to split the config in multiple files.

This also introduces a script to disable the build of packages which have a broken eval caused by broken dependencies in haskellPackages.

Our dream is that we can drastically reduce the number of packages that we need to mark as broken manually by propagating the dontDistribute flag to reverse dependencies automatically.
This has the added benefit that if a package is fixed reverse dependencies have the chance of also being fixed without human intervention.

I have a few false positives (mostly elm packages) right now which get disabled on hydra for reasons not transparent to me. Need to investigate that.

Also commit messages and code could use more documentation.

### CODE REVIEW

**Comment by maralorn** [comment]
I have the feeling that the prudent way would be to put the update of the broken-dependencies.yaml in an extra script that only branch-maintainers need to run once in a while.

**Comment by maralorn** [comment]
I have now reduced the number of false positives drastically by not marking broken packages as dont-distribute hackage2nix disables hydra for them anyway.

**Comment by maralorn** [comment]
I have the strong feeling that the remaining "false positives" are not really false positives.

They _are_ really broken and should be disabled. But since running this does not change hackage-packages.nix at all their builds were already disabled by hackage2nix for some (probably unrelated) reason.

By inspection I can‘t find a reason why hackage2nix disables them, but I think having them in this list is still informative telling us, that they are also disabled because of the broken dependency.

**Comment by maralorn** [comment]
Okay, I know this a lot to digest, but I think its in good shape now.
I will leave it as draft since the cabal2nix merge needs to happen first.

**Comment by maralorn** [comment]
btw. there is a reason that it would be nice to merge this not so far in the future. I need to painstakingly make sure that I don‘t loose any changes to configuration-hackage2nix.yaml everytime I rebase this commit on haskell-updates.

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (sternenseemann) [inline maintainers/scripts/haskell/transitive-broken-packages.nix:12]: You can implement this much simpler:

```nix
let
  brokenDeps = builtins.attrNames (
    lib.filterAttrs (
      _: v: !((builtins.tryEval (v.outPath or null)).success)
         && lib.isDerivation v
         && !v.meta.broken
    ) haskellPackages
```

The approach is to filter out all packages that fail to evaluate *despite* not being marked as broken.

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (sternenseemann) [inline maintainers/scripts/haskell/transitive-broken-packages.nix:3]: We need to pass a `config` here to make sure we don't get false positives by allowing the following things:

* `allowUnsupportedSystem`
* `allowUnfree`
* `allowInsecurePredicate = _: true;`

Not sure if I missed anything.

File: maintainers/scripts/haskell/update-hackage.sh
Comment (sternenseemann) [inline maintainers/scripts/haskell/update-hackage.sh:2]: This change seems to be left over.

**Review by sternenseemann — COMMENTED** [review]
Initial thoughts, didn't do a thorough review though.

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (sternenseemann) [reply maintainers/scripts/haskell/transitive-broken-packages.nix:3]: For unsupported system we kind of need to come up with a solution, since it involves a similar problem as marking stuff as broken unfortunately.

What I've been doing until now is adding reverse dependencies to unsupported systems. The problem is not as bad here as for marking stuff as broken, so I think we can hold off on a “proper” solution for this.

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (maralorn) [reply maintainers/scripts/haskell/transitive-broken-packages.nix:12]: Sadly I don‘t think that will work. That would filter out everything that fails to eval.
That could be anything from missing specific haskell-lib-versions to asserts in our code.

Comment (sternenseemann) [reply maintainers/scripts/haskell/transitive-broken-packages.nix:12]: We don't have any asserts in `haskellPackages`, missing attributes is not a catchable eval error in any case so both our codes would fail to eval themselves.

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (maralorn) [reply maintainers/scripts/haskell/transitive-broken-packages.nix:3]: Trying to prevent false positives seems really finnicky to me and I am afraid of overlooking something. I prefer sticking with the diff solution from above. It doesn‘t run terribly long and is conceptually simpler.
Note that I think that the code as it is does not have false positives caused by allowUnfree e.g.

Different platforms is a huge issue. I guess the output of this script actually depends on the platform we run this on. I guess in principle we could run this script for the whole matrix allowX and targetPlatform. But I wouldn‘t be sure what to do with the information. A very fine granular config file?

File: maintainers/scripts/haskell/update-hackage.sh
Comment (maralorn) [reply maintainers/scripts/haskell/update-hackage.sh:2]: What do you mean. It’s a correct fix. It’s just not super on topic for this PR. Should I put it into it’s own commit?

Comment (sternenseemann) [reply maintainers/scripts/haskell/update-hackage.sh:2]: nvm

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (sternenseemann) [reply maintainers/scripts/haskell/transitive-broken-packages.nix:3]: >  Note that I think that the code as it is does not have false positives caused by allowUnfree e.g. I think.

That is not true, it has all the issues I mentioned. It is not very noticeable since most packages support `x86_64-linux`. As an example, if you pass `localSystem = "x86_64-linux"` to your expression, it will contain (among other unsupported packages) `Euterpea` which is not broken nor are any of its dependencies.

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (maralorn) [reply maintainers/scripts/haskell/transitive-broken-packages.nix:12]: As of now there are three `assert` in configuration-common.nix.

But the point about not catchable eval errors is a good one.

File: maintainers/scripts/haskell/transitive-broken-packages.nix
Comment (maralorn) [reply maintainers/scripts/haskell/transitive-broken-packages.nix:3]: Have you tested that? I am uncertain as to why that would be happening.

We evaluate the package set twice once with allowBroken once without. When `Euterpea` doesn‘t build on x86_64-linux it wouldn‘t exist in both package sets so it shouldn‘t be in their difference either.

I may well be getting this wrong. If so I declare that I am confused and something is not working as I think it is. (Or I have written other code than I thought I wrote?)

**Comment by maralorn** [comment]
I will rebase this only when the PR is approved. But I think all voiced concerns up to now have been addressed. I talked with @sternenseemann and we agreed that the solution in the nix file is okay for now.

**Review by sternenseemann — APPROVED** [review]
I'm not sure I'm a 100% happy with the nix expression, but we can always adjust the script later — for now it seems to do its job reliably.

Merging this should be a priority since the config format is kinda integral to our workflow and sees a lot of changes.

Also we shouldn't forget to update the documentation…

**Pull request merged**
